src/tiramisu_function.cpp

```diff
@@ -116,6 +116,112 @@ isl_union_map *tiramisu::function::compute_dep_graph() {
     return result;
 }
 
+const std::map<std::string, tiramisu::buffer *> tiramisu::function::get_mapping() const
+{
+  return this->mapping;
+
+}
+
+void  tiramisu::function::add_mapping(std::pair<std::string,tiramisu::buffer *> p)
+{
+  this->mapping.insert(p);
+
+}
+
+	
+const int  &function::Automatic_communication(tiramisu::computation* c1,tiramisu::computation* c2 ) const
+{
+    std::map<std::string, tiramisu::buffer*> buff = this->get_buffers();
+    std::map<std::string, tiramisu::buffer*> mp = this->mapping;
+    std::map<std::string, tiramisu::buffer*>::iterator it ;
+    std::string name, cpt_name;
+    int i = 1;
+    char tag;  
+    tiramisu::computation* first_cpt = c1;
+    tiramisu::computation* last_cpt = c2;
+    for (it = buff.begin(); it != buff.end(); ++it)
+    {
+      name = it->second->get_name();
+      tag = it->second->tag;
+      if (it->second->get_argument_type() != tiramisu::a_temporary)
+        {
+            cpt_name= "cpt" + std::to_string(i); 
+            i++;
+            switch (tag)
+            {
+                case 'c':
+                if ((it->second->get_argument_type() == tiramisu::a_input) && (mp.find(name) != mp.end()))
+                {
+                   if (buff.find(mp.find(name)->second->get_name())->second->auto_trans== true) 
+                    {
+                        tiramisu::computation* c =  new tiramisu::computation(cpt_name,{},
+                        memcpy((*(it->second)),*(buff.find(mp.find(name)->second->get_name())->second)));
+                        (*c).then((*first_cpt),computation::root);
```

Comment:
What if the first computation has a predecessor? You can check that case and if it is the case, you should call "between" instead of "then". Or you need to check that the computation does not have a predecessor and fail otherwise.

Reply:
fist_computation is passed to this function like c1, and C1 i have got it by calling get_first_cpt , which gives me the cpt that have no predecessor in the graphe of schedule, the first_cpt is updated when i create a new cpt, by that assignment fist_cpt = comp;

the same for last_cpt; first it has the value of C2, which is the cpt that doesn't have any successor, then it is updated by the assignment last_cpt = comp , comp is the memcpy computation

Reply:
Yes I understand your code and your code is correct. What I meant by comment is that your code assumes that the input computation does not have a predecessor, so you need to add an assertion to check that (assert()). Did you notice that in Tiramisu we have plenty of assert() calls? These are useful to make sure the assumptions are valid. Because your code is valid only if these assumptions are correct, so you need to check them. Future users may use the function in a different way so you need to protect them from making a mistake in using your function.

Reply:
got it